File: feature-marker-tui/src/model/feature_state.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

/// Represents the checkpoint.json structure from state-manager.sh
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureState {
    pub version: String,
    pub feature_name: String,
    pub project_path: PathBuf,
    pub current_phase: u8,
    pub phase_status: PhaseStatus,
    pub phases: HashMap<String, Phase>,
    pub last_updated: DateTime<Utc>,
    #[serde(default)]
    pub paused: bool,
    #[serde(default)]
    pub error_state: Option<String>,
}

impl FeatureState {
    /// Create a new feature state with default values
    pub fn new(feature_name: String, project_path: PathBuf) -> Self {
        let mut phases = HashMap::new();
        phases.insert(
            "0".to_string(),
            Phase::new("Inputs Gate".to_string()),
        );
        phases.insert(
            "1".to_string(),
            Phase::new("Analysis & Planning".to_string()),
        );
        phases.insert(
            "2".to_string(),
            Phase::new("Implementation".to_string()),
        );
        phases.insert(
            "3".to_string(),
            Phase::new("Tests & Validation".to_string()),
        );
        phases.insert(
            "4".to_string(),
            Phase::new("Commit & PR".to_string()),
        );

        Self {
            version: "1.2.0".to_string(),
            feature_name,
            project_path,
            current_phase: 0,
            phase_status: PhaseStatus::Pending,
            phases,
            last_updated: Utc::now(),
            paused: false,
            error_state: None,
        }
    }

    /// Get phase by number
    pub fn get_phase(&self, phase_num: u8) -> Option<&Phase> {
        self.phases.get(&phase_num.to_string())
    }

    /// Get mutable phase by number
    pub fn get_phase_mut(&mut self, phase_num: u8) -> Option<&mut Phase> {
        self.phases.get_mut(&phase_num.to_string())
    }

    /// Check if all phases are completed
    pub fn is_completed(&self) -> bool {
        self.phases.values().all(|p| p.status == PhaseStatus::Completed)
    }

    /// Get the next pending phase
    pub fn next_pending_phase(&self) -> Option<u8> {
        for i in 0..=4 {
            if let Some(phase) = self.get_phase(i) {
                if phase.status == PhaseStatus::Pending {
                    return Some(i);
                }
            }
        }
        None
    }

    /// Get completion percentage
    pub fn completion_percentage(&self) -> u8 {
        let completed = self
            .phases
            .values()
            .filter(|p| p.status == PhaseStatus::Completed)
            .count();
        ((completed * 100) / self.phases.len()) as u8
    }
}

/// Represents a single phase in the workflow
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Phase {
    pub name: String,
    pub status: PhaseStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub started_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_task_index: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tasks: Option<u32>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub completed_tasks: Vec<u32>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub outputs: Vec<String>,
}

impl Phase {
    pub fn new(name: String) -> Self {
        Self {
            name,
            status: PhaseStatus::Pending,
            started_at: None,
            completed_at: None,
            current_task_index: None,
            total_tasks: None,
            completed_tasks: Vec::new(),
            outputs: Vec::new(),
        }
    }

    /// Start the phase
    pub fn start(&mut self) {
        self.status = PhaseStatus::InProgress;
        self.started_at = Some(Utc::now());
    }

    /// Complete the phase
    pub fn complete(&mut self) {
        self.status = PhaseStatus::Completed;
        self.completed_at = Some(Utc::now());
    }

    /// Mark phase as error
    pub fn error(&mut self) {
        self.status = PhaseStatus::Error;
    }

    /// Get task progress as (current, total)
    pub fn task_progress(&self) -> Option<(u32, u32)> {
        match (self.current_task_index, self.total_tasks) {
            (Some(current), Some(total)) => Some((current, total)),
            _ => None,
        }
    }
}

/// Status of a phase
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum PhaseStatus {
    #[default]
    Pending,
    InProgress,
    Completed,
    Error,
}

impl PhaseStatus {
    /// Get display symbol for the status
    pub fn symbol(&self) -> &'static str {
        match self {
            PhaseStatus::Pending => "○",
            PhaseStatus::InProgress => "→",
            PhaseStatus::Completed => "✓",
            PhaseStatus::Error => "✗",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_new_feature_state() {
        let state = FeatureState::new(
            "test-feature".to_string(),
            PathBuf::from("/tmp/project"),
        );

        assert_eq!(state.feature_name, "test-feature");
        assert_eq!(state.current_phase, 0);
        assert_eq!(state.phases.len(), 5);
        assert!(!state.paused);
    }

    #[test]
    fn test_phase_progress() {
        let mut phase = Phase::new("Test".to_string());
        phase.current_task_index = Some(3);
        phase.total_tasks = Some(10);

        assert_eq!(phase.task_progress(), Some((3, 10)));
    }

    #[test]
    fn test_completion_percentage() {
        let mut state = FeatureState::new(
            "test".to_string(),
            PathBuf::from("/tmp"),
        );

        // Complete 2 out of 5 phases
        state.get_phase_mut(0).unwrap().complete();
        state.get_phase_mut(1).unwrap().complete();

        assert_eq!(state.completion_percentage(), 40);
    }
}

File: feature-marker-menubar/Sources/FeatureMarkerMenuBar/Models/FeatureState.swift
import Foundation

// MARK: - Phase Status

enum PhaseStatus: String, Codable, Sendable {
    case pending
    case inProgress = "in_progress"
    case completed
    case error
}

// MARK: - Phase

struct Phase: Codable, Sendable {
    let name: String
    let status: PhaseStatus
    var startedAt: Date?
    var completedAt: Date?
    var currentTaskIndex: Int?
    var totalTasks: Int?
    var completedTasks: [Int] = []
    var outputs: [String] = []

    enum CodingKeys: String, CodingKey {
        case name, status
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case currentTaskIndex = "current_task_index"
        case totalTasks = "total_tasks"
        case completedTasks = "completed_tasks"
        case outputs
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = try container.decode(String.self, forKey: .name)
        status = try container.decode(PhaseStatus.self, forKey: .status)
        startedAt = try container.decodeIfPresent(Date.self, forKey: .startedAt)
        completedAt = try container.decodeIfPresent(Date.self, forKey: .completedAt)
        currentTaskIndex = try container.decodeIfPresent(Int.self, forKey: .currentTaskIndex)
        totalTasks = try container.decodeIfPresent(Int.self, forKey: .totalTasks)
        completedTasks = (try? container.decode([Int].self, forKey: .completedTasks)) ?? []
        outputs = (try? container.decode([String].self, forKey: .outputs)) ?? []
    }
}

// MARK: - Feature State (checkpoint.json)

struct FeatureState: Codable, Sendable {
    let version: String
    let featureName: String
    let projectPath: String
    let currentPhase: Int
    let phaseStatus: PhaseStatus
    let phases: [String: Phase]
    let lastUpdated: Date
    var paused: Bool
    var errorState: String?

    enum CodingKeys: String, CodingKey {
        case version
        case featureName = "feature_name"
        case projectPath = "project_path"
        case currentPhase = "current_phase"
        case phaseStatus = "phase_status"
        case phases
        case lastUpdated = "last_updated"
        case paused
        case errorState = "error_state"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        version = try container.decode(String.self, forKey: .version)
        featureName = try container.decode(String.self, forKey: .featureName)
        projectPath = try container.decode(String.self, forKey: .projectPath)
        currentPhase = try container.decode(Int.self, forKey: .currentPhase)
        phaseStatus = try container.decode(PhaseStatus.self, forKey: .phaseStatus)
        phases = try container.decode([String: Phase].self, forKey: .phases)
        lastUpdated = try container.decode(Date.self, forKey: .lastUpdated)
        paused = (try? container.decode(Bool.self, forKey: .paused)) ?? false
        errorState = try container.decodeIfPresent(String.self, forKey: .errorState)
    }
}

// MARK: - Feature Summary (for list display)

struct FeatureSummary: Identifiable, Sendable {
    var id: String { name }
    let name: String
    let currentPhase: Int
    let totalPhases: Int
    let phaseStatus: PhaseStatus
    let lastUpdated: Date
    let isRunning: Bool
    let paused: Bool
    let hasError: Bool

    var progress: Double {
        Double(currentPhase) / Double(totalPhases)
    }
}

// MARK: - Running Feature

struct RunningFeature: Sendable {
    let name: String
    let mode: String
    let startedAt: Date
    var processID: Int32?
}

// MARK: - Output Line

struct OutputLine: Identifiable, Sendable {
    let id = UUID()
    let text: String
    let isStderr: Bool
    let timestamp = Date()

    var isSuccess: Bool {
        text.contains("\u{2713}") || text.contains("\u{2705}")
    }
}

// MARK: - JSON Decoder

extension JSONDecoder {
    static var checkpoint: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)

            // ISO 8601 with fractional seconds (microseconds)
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = formatter.date(from: dateString) {
                return date
            }

            // Fallback: without fractional seconds
            formatter.formatOptions = [.withInternetDateTime]
            if let date = formatter.date(from: dateString) {
                return date
            }

            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Cannot decode date: \(dateString)"
            )
        }
        return decoder
    }
}

// MARK: - Phase Names

enum PhaseInfo {
    static let names = [
        "Inputs Gate",
        "Analysis & Planning",
        "Implementation",
        "Tests & Validation",
        "Commit & PR"
    ]
    static let count = 5
}
